### ISSUE TITLE
Logging a null value causes the app to crash with a stack overflow

### ISSUE THREAD

**Issue by alwilton** [serilog/serilog#1400]
If a null value is detected and logged the app crashes with a stack overflow - any other logging is fine, only a log entry after a null detection is an issue. If the log line is removed the app behaves as expected. The app is a .net core 3.1 web api app using the following relevant packages:
```
    <PackageReference Include="Serilog.AspNetCore" Version="3.2.0" />
    <PackageReference Include="Serilog.Settings.Configuration" Version="3.1.1-dev-00216" />
    <PackageReference Include="Serilog.Sinks.Console" Version="4.0.0-dev-00837" />
    <PackageReference Include="Serilog.Sinks.Email" Version="2.3.1-dev-00101" />
    <PackageReference Include="Serilog.Sinks.File" Version="4.1.0" />
```
the config file is as shown in the this [link ](https://stackoverflow.com/questions/50387469/net-core-and-serilog-email-sink-json-config/59914193#59914193)

Calling the following code results in a stack overflow crash
```
using System;
using System.Reflection;
using Microsoft.AspNetCore.Mvc;
using Serilog;

namespace TrackumApi.Controllers.Lab
{
    [Route("api/[controller]")]
    [ApiController]
    public class TestController : ControllerBase
    {
        [HttpGet("TryCrash")]
        public string TryCrash()
        {
            var m = MethodBase.GetCurrentMethod();
            try
            {
                var sr = GetFile("123");
                return sr;
            }
            catch (Exception ex)
            {
                Log.Error(m.ReflectedType != null
                    ? $"{m.ReflectedType.Name} {m.Name} : {ex.Message}"
                    : $"{m.Name} : {ex.Message}");
            }

            return null;
        }

        private string GetFile(string fileno)
        {
            var m = MethodBase.GetCurrentMethod();
            try
            {
                /*
                 * WARNING - logging any null value results in a stack overflow and crash.
                 */
                object file = null;
                if (file == null)
                {
                    Log.Warning($"File {fileno} file is null"); // <== after this log is written the app crashes with a stack overflow
                    return "Fileno is null";
                }

                return "test OK";
            }
            catch (Exception e)
            {
                Log.Error(m.ReflectedType != null
                    ? $"{m.ReflectedType.Name} {m.Name} {e.Message}"
                    : $"{m.Name} {e.Message} - {fileno}");
            }

            return null;
        }
    }
}
```

```
[11:09:55.313 [Warning]  File 123 file is null
Stack overflow.
```
SelfLog is enabled to the console but nothing appears.

Let me know if anything more is needed. TIA

**Comment by nblumhardt** [comment]
It looks like the cause of the crash is only tangentially related to whether `file` is null or not - the `$` in front of `$"File {fileno} file is null"` uses string interpolation, and therefore the `null` is formatted into the string before it ever hits Serilog. (You should [leave out the `$` and use message templates](https://nblumhardt.com/2015/01/c-6-string-interpolation-and-serilog/) to get all the structured logging benefits of Serilog.)

In the case that `file` is non-null, are you doing any additional work that might prevent a stack overflow down the line, or are you using `file = new object()`? If not the latter, then this is worth a try to narrow things down.

To properly debug this we'll need the stack trace attached to the `StackOverflowException` - you should be able to get this with the VS or Rider debugger.

HTH!

**Comment by alwilton** [comment]
Thx – to be clear “file” is set to null in this contrived example but “fileno” is a valid string “123” so the $"File {fileno} file is null”  should not be affected and indeed it does emit to the console, it is after that the SO occurs, but I’ll remove it from the sample. I’ll try to get the stack trace later tonight.

-Al

**Comment by alwilton** [comment]
After a very long night I've found the solution. It had nothing to do with a null value. It was the "Warning" level which brought the Email extension into play and that was the culprit. First, and the big one is that calling functions from the configurations created an email extension for each function call. I don't understand why but it does. So I read and prepared all email extension variables before the extension was added as a "WriteTo" item. That got me halfway there. Leaving the "WriteTo" custom email section in the SeriLog appsetting section created an Email sink and then the "WriteTo.EmailCustom" created another and so I needed to move the "WriteTo.EmailCustom" section out of the "SeriLog" section into its own section so only one email sink was created. It now seems to be working and stable. I will need to update my StackOverflow post on the custom email extension tomorrow. If you care to look at the current SO post at [https://stackoverflow.com/questions/50387469/net-core-and-serilog-email-sink-json-config/59914193#59914193](https://stackoverflow.com/questions/50387469/net-core-and-serilog-email-sink-json-config/59914193#59914193) and explain why every call to a function in configuration creates another email sink that would demystify things a bit. Thx.
